packages/utils/src/fallbackSubgraph.ts

```diff
@@ -0,0 +1,80 @@
+import { Logger } from "./logger";
+import { createLoggingContext } from "./request";
+
+export type SubgraphSyncRecord = {
+  synced: boolean;
+  latestBlock: number;
+  syncedBlock: number;
+}
+
+type Sdk<T> = {
+  client: T;
+  record: SubgraphSyncRecord;
+};
+
+export class FallbackSubgraph<T extends { GetBlockNumber: () => Promise<any>; }> {
+  private readonly sdks: Sdk<T>[];
+
+  private get synced(): Sdk<T>[] {
+    return this.sdks.filter((sdk) => sdk.record.synced);
+  }
+
+  private get outOfSync(): Sdk<T>[] {
+    return this.sdks.filter((sdk) => !sdk.record.synced);
+  }
+
+  constructor(
+    private readonly logger: Logger,
+    private readonly chainId: number,
+    sdks: T[],
+    // Subgraph is considered out of sync if it is more than this many blocks behind the latest block.
+    private readonly subgraphSyncBuffer: number,
+  ) {
+    this.sdks = sdks.map((client) => ({
+      client,
+      record: {
+        synced: false,
+        latestBlock: 0,
+        syncedBlock: 0,
+      },
+    }));
+  }
+
+  public getSynced(): T {
+    const synced = this.synced;
+    if (synced.length === 0) {
+      throw new Error("No subgraphs available / in-sync!");
+    }
+    // Get a random client to ensure we are varying which subgraph client we're using.
+    return synced[Math.floor(Math.random() * synced.length)].client;
+  }
```

Comment:
why not preferentially get the most in sync client? then select randomly if there are multiple clients that are at the same processing block

Reply:
sure, i like that

Reply:
ohh woops, I implemented a weighted system, must have misremembered this comment lol.

Should I go with the weighted system, preferring *more* in sync..? Hmmm, I think actually the way you described it would be best